rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -455,110 +542,354 @@ where
             dataset_definition,
-            dataset_data_transformer,
-            reporter,
+            asset_status,
+            asset_specification,
+            device_specification,
+            forwarder,
+            connector_context,
         }
     }
-    /// Used to report the status and/or [`MessageSchema`] of an dataset
+
+    /// Used to report the status of a dataset
+    pub async fn report_status(&self, status: Result<(), ConfigError>) {
+        let adr_asset_status = azure_device_registry::AssetStatus {
+            // TODO: Do I need to include the version here?
+            // config: Some(azure_device_registry::StatusConfig {
+            //     version: self.asset_specification.version,
+            //     ..azure_device_registry::StatusConfig::default()
+            // }),
+            datasets: Some(vec![azure_device_registry::DatasetEventStreamStatus {
+                name: self.dataset_ref.dataset_name.clone(),
+                message_schema_reference: None,
+                error: status.err(),
+            }]),
+            ..azure_device_registry::AssetStatus::default()
+        };
+
+        // send status update to the service
+        AssetClient::internal_report_status(
+            adr_asset_status,
+            &self.connector_context,
+            &self.asset_ref,
+            &self.asset_status,
+        )
+        .await;
+    }
+
+    /// Used to report the message schema of a dataset
+    ///
     /// # Errors
-    /// TODO
-    pub fn report_status(
+    /// [`schema_registry::Error`] of kind [`InvalidArgument`](schema_registry::ErrorKind::InvalidArgument)
+    /// if the content of the [`MessageSchema`] is empty or there is an error building the request
+    ///
+    /// [`schema_registry::Error`] of kind [`ServiceError`](schema_registry::ErrorKind::ServiceError)
+    /// if there is an error returned by the Schema Registry Service.
+    ///
+    /// # Panics
+    /// If the Schema Registry Service returns a schema without required values. This should get updated
+    /// to be validated by the Schema Registry API surface in the future
+    pub async fn report_message_schema(
         &self,
-        status: Result<Option<MessageSchema>, ConfigError>,
-    ) -> Result<Option<MessageSchemaReference>, String> {
-        // Report the status of the dataset
-        self.reporter.report_status(status)
+        message_schema: MessageSchema,
+    ) -> Result<MessageSchemaReference, schema_registry::Error> {
+        // TODO: save message schema provided with message schema uri so it can be compared
+        // send message schema to schema registry service
+        let message_schema_reference = Retry::spawn(
+            RETRY_STRATEGY,
+            async || -> Result<schema_registry::Schema, RetryError<schema_registry::Error>> {
+                self.connector_context
+                    .schema_registry_client
+                    .put(
+                        message_schema.clone(),
```

Comment:
Does this have to be cloned? I see that it doesn't get used anywhere else in the function and it is not a reference

Reply:
It needs to be there because of the retry :( A bit annoying, but because of the retry, all values need to be able to be used again, similar to if they're in a loop

Reply:
Is this going to be an issue if the schema gets updated somehow?

Reply:
this is where the schema gets updated, so it would be pretty weird to call it twice concurrently. There is some future work around when retries should stop retrying though.